### PR TITLE
Add path constraints to Files, and support for more containers.

### PR DESCRIPTION
**Pull request by ttaa9**
Attempts to satisfy the requirements of #51 and #58. 

If the schema changes seem ok, let me know and I will update the tools (and cbrain) as well.

### CODE REVIEW

File: schema/descriptor.schema.json [hunk 1]
```diff
@@ -28,18 +28,57 @@
...
+                        },
+                        "index": {
+                            "id": "http://github.com/boutiques/boutiques-schema/container/index",
+                            "minLength": 1,
+                            "description": "Index where the image is available.",
+                            "default": "http://index.docker.io",
+                            "type": "string"
+                        }
+                    },
+                    "required": ["type", "image", "index"]
```
Comment (glatard) [inline schema/descriptor.schema.json]: I'm not sure if `index` should be required since there is a default value for it.

File: schema/descriptor.schema.json [hunk 2]
```diff
@@ -28,18 +28,57 @@
...
-            "id": "http://github.com/boutiques/boutiques-schema/docker-index",
-            "minLength": 1,
-            "description": "Docker index where Docker image is available.",
-            "default": "http://index.docker.io",
-            "type": "string"
+        "container": {
+            "id": "http://github.com/boutiques/boutiques-schema/container",
+            "type": "object",
+            "description": "Object representing a container for the tool.",
+            "allOf": [{
```
Comment (glatard) [inline schema/descriptor.schema.json]: I don't understand why `allOf` is required here.

File: schema/descriptor.schema.json [hunk 3]
```diff
@@ -28,18 +28,57 @@
...
+                "properties": {
+                    "launch-directory": {
```
Comment (glatard) [inline schema/descriptor.schema.json]: I would call it `working-directory` to be consistent with the terminology used in `Docker`.

File: schema/descriptor.schema.json (hunk 2 above)
Comment (ttaa9) [inline schema/descriptor.schema.json]: It allows potentially letting `container-hash` and `launch-directory` coexist with either the `docker` spec or the `rootfs/singularity` spec. It does not force them to be required. It just lets both possible specs match to these so that the same values dont have to be written out twice (i.e. once in each of the two container type specs).

File: schema/descriptor.schema.json (hunk 1 above)
Comment (ttaa9) [inline schema/descriptor.schema.json]: Fair enough

File: schema/descriptor.schema.json (hunk 3 above)
Comment (ttaa9) [inline schema/descriptor.schema.json]: Sure

File: schema/descriptor.schema.json (hunk 2 above)
Comment (glatard) [inline schema/descriptor.schema.json]: OK

Comment (glatard) [inline schema/descriptor.schema.json]: Well, isn't it possible to have a container object without an implementation specified then? For instance an empty object?

Comment (glatard) [inline schema/descriptor.schema.json]: -> it's not possible thanks to the `required` property in *both* docker and rootfs/singularity sections, right?

Comment (ttaa9) [inline schema/descriptor.schema.json]: That's correct; it has to satisfy exactly one of (`oneOf`) those two schemas, and each has its own `required` properties. Testing with an empty container gives an error because of this.

**Comment by glatard** [comment]
@kofalt @chrisfilo @johnflavin would you like to review this PR too?

**Comment by kofalt** [comment]
@glatard Totally! I'll see if I can give it a look today, but it might end up being tomorrow morning :smile:

**Comment by chrisfilo** [comment]
I'm not trying to be picky, but isn't "image" a more suitable name than "container"?

File: schema/descriptor.schema.json [hunk 4]
```diff
@@ -257,25 +296,50 @@
...
                         "description": "Specify the maximum number of entries in the list. May only be used with List type inputs.",
                         "type": "number"
+                    },
+                    "absolute-path": {
+                        "id": "http://github.com/boutiques/boutiques-schema/input/absolute-path",
+                        "description": "Specifies that this input must be given as an absolute path. Only specifiable for File type inputs. Cannot be true if relative-path is true.",
+                        "type": "boolean"
+                    },
+                    "relative-path": {
+                        "id": "http://github.com/boutiques/boutiques-schema/input/relative-path",
```
Comment (johnflavin) [inline schema/descriptor.schema.json]: Is it necessary to have two mutually-exclusive boolean properties? If you set one to `true`, the other must be `false`. So maybe there could just be one input.

On the other hand, there is the middle case where the image doesn't care about whether the paths are absolute or relative, so neither property would be true. That is a benefit to having two properties. But in that case the tool implementation would have to decide at launch time to either make the path relative or absolute; it wouldn’t matter which one but it does have to set one or the other. But we could guide the tool writers by setting one of the two properties `true` by default.

I guess what I would want is for there to be one boolean property. If `true`, then the paths must be absolute; if `false`, then relative. (Or the other way. Whatever.) And one of those (absolute?) should be the default value if nothing is specified. No guess work by the tool developer.

File: tools/validator.rb [hunk 5]
```diff
@@ -49,10 +49,11 @@ def usage
 ### Validation of descriptor arguments ###
 
 ## Helper functions ##
-inputGet  = lambda { |s| descriptor['inputs'].map {       |v| v[s] } rescue [] }
-outputGet = lambda { |s| descriptor['output-files'].map { |v| v[s] } rescue [] }
-groupGet  = lambda { |s| descriptor['groups'].map {       |v| v[s] } rescue [] }
-inById    = lambda { |i| descriptor['inputs'].find{       |v| v['id']==i } || {} }
+safeGet   = lambda { |sec,targ| descriptor[sec].map { |v| v[targ] }.compact rescue [] }
```
Comment (johnflavin) [inline tools/validator.rb]: Is this stuff in scope of this PR?

**Comment by glatard** [comment]
@chrisfilo you're right, how about `container-image`?

File: tools/validator.rb (hunk 5 above)
Comment (glatard) [inline tools/validator.rb]: Yes you're right, small fix though, let's keep it here if you don't mind.

**Comment by chrisfilo** [comment]
+1

File: schema/descriptor.schema.json (hunk 4 above)
Comment (glatard) [inline schema/descriptor.schema.json]: The idea was that if none of these properties are here, then it is up to the implementation to decide which one to use. There might be good reasons to use one or the other. In CBRAIN for instance, relative paths are always preferred because it makes the tasks relocatable in different clusters (which is not really an issue with containers but still is the preferred way). 

But I think I'm fine with a single `use-absolute-path` boolean property, the default being to use relative paths.

File: schema/descriptor.schema.json [hunk 6]
```diff
@@ -28,18 +28,57 @@
...
+                        "type": "string"
+                    },
+                    "container-hash": {
+                        "id": "http://github.com/boutiques/boutiques-schema/container/container-hash",
+                        "minLength": 1,
+                        "description": "Hash for the given container.",
+                        "type": "string"
+                    }
+                }
+            },{
```
Comment (kofalt) [inline schema/descriptor.schema.json]: Related to the earlier conversation: couldn't you move this `properties` key up to the `container` level? Then you could ditch the `allOf` -> `oneOf` nesting (you'd just need `oneOf`).

File: schema/descriptor.schema.json [hunk 7]
```diff
@@ -28,18 +28,57 @@
...
+                        "description": "Location from which this task must be launched within the container.",
+                        "type": "string"
+                    },
+                    "container-hash": {
+                        "id": "http://github.com/boutiques/boutiques-schema/container/container-hash",
+                        "minLength": 1,
+                        "description": "Hash for the given container.",
+                        "type": "string"
+                    }
+                }
```
Comment (kofalt) [inline schema/descriptor.schema.json]: Related to the earlier conversation: couldn't you move this `properties` key up to the `container` level? Then you could ditch the `allOf` -> `oneOf` nesting (you'd just need `oneOf`).

File: schema/descriptor.schema.json [hunk 8]
```diff
@@ -28,18 +28,57 @@
...
+                        },
+                        "index": {
+                            "id": "http://github.com/boutiques/boutiques-schema/container/index",
+                            "minLength": 1,
+                            "description": "Index where the image is available.",
+                            "default": "http://index.docker.io",
+                            "type": "string"
+                        }
+                    },
+                    "required": ["type", "image"]
```
Comment (kofalt) [inline schema/descriptor.schema.json]: Add a `description` key here talking about what kind of container this is (docker)

File: schema/descriptor.schema.json [hunk 9]
```diff
@@ -28,18 +28,57 @@
...
+                    "properties": {
+                        "type": { "enum": ["rootfs", "singularity"] },
+                        "url": {
+                            "id": "http://github.com/boutiques/boutiques-schema/container/url",
+                            "minLength": 1,
+                            "description": "URL where the image is available.",
+                            "type": "string"
+                        }
+                    },
+                    "required": ["type", "url"]
```
Comment (kofalt) [inline schema/descriptor.schema.json]: Add a `description` key here talking about what kind of container this is (singularity or rootfs)

**Comment by kofalt** [comment]
Review complete.

I'm not a fan of the absolute / relative feature here - I'm wondering if we can simplify things a bit?

I know of tools that require an absolute path, but I don't know of tools that *require* a relative path. Do you think we could get away with having just a boolean such as `absolutePathRequired`? If true, path must be absolute, if false, anything goes?

Motivation is that, like @glatard's comment on #51, I am thinking about the complexity of the command-generation bit and it would be nice to keep that simpler wherever possible. There's also something to be said for keeping the schema from getting too complex...

File: schema/descriptor.schema.json (hunk 4 above)
Comment (ttaa9) [inline schema/descriptor.schema.json]: I prefer a single `use-absolute-path` as well. 

One minor thing is the semantics of this property: does it mean that an input must be absolute (i.e. it just checks it) vs the platform must transform it, or is the interpretation up to the platform?

File: schema/descriptor.schema.json (hunk 7 above)
Comment (ttaa9) [inline schema/descriptor.schema.json]: I greatly like that idea, but I don't immediately see a clean way to do it, because (with the current names) the `required` property depends on which `oneOf` is being matched, and `required` must be on the same level as `properties`. (If you know of one, please let me know!)

File: schema/descriptor.schema.json (hunk 4 above)
Comment (johnflavin) [inline schema/descriptor.schema.json]: My interpretation is that if an input specifies `use-absolute-path=true`, then the platform must give that input an absolute-path value before it attempts to launch the container. The contract implied by this property is between the platform and the container. 

The contract between platform and user is another matter, and not covered by this property. The platform could require the user to give absolute paths to inputs with this property, or could translate user-provided relative paths to absolute. Whichever.

File: schema/descriptor.schema.json (hunk 7 above)
Comment (glatard) [inline schema/descriptor.schema.json]: @kofalt any suggestion on this?
